File: docs/beginner/module-04-simple-queries/code-samples/csharp/05_AdvancedQueries.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Azure;
using Azure.Search.Documents;
using Azure.Search.Documents.Models;
using Microsoft.Extensions.Configuration;

namespace AzureSearchModule4
{
    /// <summary>
    /// Module 4: Simple Queries and Filters - Advanced Queries
    /// 
    /// This class demonstrates advanced query techniques in Azure AI Search using C#.
    /// Learn about field boosting, fuzzy search, wildcards, and complex query patterns.
    /// 
    /// Prerequisites:
    /// - Azure AI Search service configured
    /// - Sample index with data (from previous modules)
    /// - Configuration set up in appsettings.json or environment variables
    /// 
    /// Author: Azure AI Search Tutorial
    /// </summary>
    public class AdvancedQueries
    {
        private readonly SearchClient _searchClient;
        private readonly IConfiguration _configuration;

        public AdvancedQueries()
        {
            // Load configuration
            _configuration = new ConfigurationBuilder()
                .AddJsonFile("appsettings.json", optional: true)
                .AddEnvironmentVariables()
                .Build();

            // Initialize search client
            _searchClient = CreateSearchClient();
        }

        /// <summary>
        /// Create and return an Azure AI Search client.
        /// </summary>
        /// <returns>Configured SearchClient</returns>
        private SearchClient CreateSearchClient()
        {
            var serviceEndpoint = _configuration["AzureSearch:ServiceEndpoint"] ?? 
                                 Environment.GetEnvironmentVariable("AZURE_SEARCH_SERVICE_ENDPOINT");
            var apiKey = _configuration["AzureSearch:ApiKey"] ?? 
                        Environment.GetEnvironmentVariable("AZURE_SEARCH_API_KEY");
            var indexName = _configuration["AzureSearch:IndexName"] ?? 
                           Environment.GetEnvironmentVariable("AZURE_SEARCH_INDEX_NAME");

            if (string.IsNullOrEmpty(serviceEndpoint) || string.IsNullOrEmpty(apiKey) || string.IsNullOrEmpty(indexName))
            {
                throw new InvalidOperationException("Missing required configuration. Check your appsettings.json or environment variables.");
            }

            var serviceUri = new Uri(serviceEndpoint);
            var credential = new AzureKeyCredential(apiKey);

            return new SearchClient(serviceUri, indexName, credential);
        }

        /// <summary>
        /// Display search results in a formatted way.
        /// </summary>
        /// <param name="results">Search results</param>
        /// <param name="title">Title for the result set</param>
        /// <param name="maxResults">Maximum number of results to display</param>
        public static void DisplayResults(SearchResults<SearchDocument> results, string title, int maxResults = 5)
        {
            Console.WriteLine($"\n{new string('=', 60)}");
            Console.WriteLine(title);
            Console.WriteLine(new string('=', 60));

            var resultList = results.GetResults().Take(maxResults).ToList();

            if (!resultList.Any())
            {
                Console.WriteLine("No results found.");
                return;
            }

            for (int i = 0; i < resultList.Count; i++)
            {
                var result = resultList[i];
                var document = result.Document;

                Console.WriteLine($"\n{i + 1}. {document.GetValueOrDefault("title", "No title")}");
                Console.WriteLine($"   Score: {result.Score:F4}");
                Console.WriteLine($"   Category: {document.GetValueOrDefault("category", "N/A")}");
                Console.WriteLine($"   Rating: {document.GetValueOrDefault("rating", "N/A")}");

                // Show content preview
                var content = document.GetValueOrDefault("content", "")?.ToString();
                if (!string.IsNullOrEmpty(content))
                {
                    var preview = content.Length > 100 ? content.Substring(0, 100) + "..." : content;
                    Console.WriteLine($"   Preview: {preview}");
                }
            }

            if (results.GetResults().Count() > maxResults)
            {
                Console.WriteLine($"\n... and {results.GetResults().Count() - maxResults} more results");
            }
        }

        /// <summary>
        /// Demonstrate field boosting to influence relevance scoring.
        /// </summary>
        public async Task FieldBoostingExamplesAsync()
        {
            Console.WriteLine("\n" + new string('=', 80));
            Console.WriteLine("FIELD BOOSTING EXAMPLES");
            Console.WriteLine(new string('=', 80));

            // Example 1: No boosting (baseline)
            Console.WriteLine("\n1. No Field Boosting (Baseline)");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions
                {
                    Size = 5
                };
                searchOptions.SearchFields.Add("title");
                searchOptions.SearchFields.Add("content");

                var results = await _searchClient.SearchAsync<SearchDocument>("python tutorial", searchOptions);
                DisplayResults(results.Value, "No boosting - equal weight for all fields");
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }

            // Example 2: Boost title field
            Console.WriteLine("\n2. Boost Title Field (3x weight)");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions
                {
                    Size = 5
                };
                searchOptions.SearchFields.Add("title^3"); // Title weighted 3x
                searchOptions.SearchFields.Add("content");

                var results = await _searchClient.SearchAsync<SearchDocument>("python tutorial", searchOptions);
                DisplayResults(results.Value, "Title boosted 3x - title matches score higher");
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }

            // Example 3: Multiple field boosting
            Console.WriteLine("\n3. Multiple Field Boosting");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions
                {
                    Size = 5
                };
                searchOptions.SearchFields.Add("title^5"); // Title 5x
                searchOptions.SearchFields.Add("category^2"); // Category 2x
                searchOptions.SearchFields.Add("content"); // Content 1x

                var results = await _searchClient.SearchAsync<SearchDocument>("azure machine learning", searchOptions);
                DisplayResults(results.Value, "Title 5x, category 2x, content 1x");
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }
        }

        /// <summary>
        /// Demonstrate fuzzy search for handling typos and variations.
        /// </summary>
        public async Task FuzzySearchExamplesAsync()
        {
            Console.WriteLine("\n" + new string('=', 80));
            Console.WriteLine("FUZZY SEARCH EXAMPLES");
            Console.WriteLine(new string('=', 80));

            // Example 1: Exact match (baseline)
            Console.WriteLine("\n1. Exact Match (Baseline)");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions
                {
                    QueryType = SearchQueryType.Simple,
                    Size = 3
                };

                var results = await _searchClient.SearchAsync<SearchDocument>("machine", searchOptions);
                DisplayResults(results.Value, "Exact match: 'machine'");
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }

            // Example 2: Fuzzy search with typo
            Console.WriteLine("\n2. Fuzzy Search with Typo");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions
                {
                    QueryType = SearchQueryType.Full, // Requires full Lucene syntax
                    Size = 3
                };

                var results = await _searchClient.SearchAsync<SearchDocument>("machne~", searchOptions); // Typo: missing 'i'
                DisplayResults(results.Value, "Fuzzy search: 'machne~' (should match 'machine')");
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }

            // Example 3: Fuzzy search with edit distance
            Console.WriteLine("\n3. Fuzzy Search with Edit Distance");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions
                {
                    QueryType = SearchQueryType.Full,
                    Size = 3
                };

                var results = await _searchClient.SearchAsync<SearchDocument>("learing~2", searchOptions); // Allow up to 2 character differences
                DisplayResults(results.Value, "Fuzzy search: 'learing~2' (should match 'learning')");
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }
        }

        /// <summary>
        /// Demonstrate wildcard search patterns.
        /// </summary>
        public async Task WildcardSearchExamplesAsync()
        {
            Console.WriteLine("\n" + new string('=', 80));
            Console.WriteLine("WILDCARD SEARCH EXAMPLES");
            Console.WriteLine(new string('=', 80));

            // Example 1: Suffix wildcard
            Console.WriteLine("\n1. Suffix Wildcard");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions { Size = 5 };
                var results = await _searchClient.SearchAsync<SearchDocument>("develop*", searchOptions); // Matches develop, developer, development, etc.
                DisplayResults(results.Value, "Suffix wildcard: 'develop*'");
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }

            // Example 2: Prefix wildcard (requires full Lucene)
            Console.WriteLine("\n2. Prefix Wildcard");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions
                {
                    QueryType = SearchQueryType.Full,
                    Size = 5
                };

                var results = await _searchClient.SearchAsync<SearchDocument>("*ing", searchOptions); // Matches words ending in 'ing'
                DisplayResults(results.Value, "Prefix wildcard: '*ing'");
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }

            // Example 3: Middle wildcard
            Console.WriteLine("\n3. Middle Wildcard");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions
                {
                    QueryType = SearchQueryType.Full,
                    Size = 5
                };

                var results = await _searchClient.SearchAsync<SearchDocument>("mach*ne", searchOptions); // Matches machine, etc.
                DisplayResults(results.Value, "Middle wildcard: 'mach*ne'");
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }
        }

        /// <summary>
        /// Demonstrate complex query combinations.
        /// </summary>
        public async Task ComplexQueryCombinationsAsync()
        {
            Console.WriteLine("\n" + new string('=', 80));
            Console.WriteLine("COMPLEX QUERY COMBINATIONS");
            Console.WriteLine(new string('=', 80));

            // Example 1: Boosting + Fuzzy + Filter
            Console.WriteLine("\n1. Boosting + Fuzzy + Filter");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions
                {
                    QueryType = SearchQueryType.Full,
                    Filter = "rating ge 3.0",
                    Size = 3
                };
                searchOptions.SearchFields.Add("title^3");
                searchOptions.SearchFields.Add("content");

                var results = await _searchClient.SearchAsync<SearchDocument>("machne~ learing~", searchOptions); // Fuzzy search
                DisplayResults(results.Value, "Fuzzy + boosting + filter combination");
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }

            // Example 2: Wildcard + Filter + Sorting
            Console.WriteLine("\n2. Wildcard + Filter + Sorting");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions
                {
                    QueryType = SearchQueryType.Full,
                    Filter = "category eq 'Technology'",
                    Size = 3
                };
                searchOptions.OrderBy.Add("rating desc");

                var results = await _searchClient.SearchAsync<SearchDocument>("develop* AND tutorial", searchOptions); // Wildcard + AND
                DisplayResults(results.Value, "Wildcard + filter + sorting");
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }

            // Example 3: Multiple techniques with highlighting
            Console.WriteLine("\n3. Multiple Techniques + Highlighting");
            Console.WriteLine(new string('-', 40));

            try
            {
                var searchOptions = new SearchOptions
                {
                    QueryType = SearchQueryType.Full,
                    Filter = "category eq 'Technology'",
                    Size = 3
                };
                searchOptions.SearchFields.Add("title^2");
                searchOptions.SearchFields.Add("content");
                searchOptions.HighlightFields.Add("title");
                searchOptions.HighlightFields.Add("content");
                searchOptions.OrderBy.Add("@search.score desc");

                var results = await _searchClient.SearchAsync<SearchDocument>("python* OR java*", searchOptions); // Wildcard OR
                DisplayResults(results.Value, "Complex combination with highlighting");

                // Show highlights
                var resultList = results.Value.GetResults().Take(2).ToList();
                for (int i = 0; i < resultList.Count; i++)
                {
                    var result = resultList[i];
                    if (result.Document.ContainsKey("@search.highlights"))
                    {
                        Console.WriteLine($"\nHighlights for result {i + 1}:");
                        var highlights = result.Document["@search.highlights"] as IDictionary<string, object>;
                        if (highlights != null)
                        {
                            foreach (var highlight in highlights.Take(2))
                            {
                                if (highlight.Value is IEnumerable<object> highlightList)
                                {
                                    foreach (var item in highlightList.Take(1))
                                    {
                                        Console.WriteLine($"  {highlight.Key}: {item}");
                                    }
                                }
                            }
                        }
                    }
                }
            }
            catch (RequestFailedException ex)
            {
                Console.WriteLine($"Search failed: {ex.Message}");
            }
        }

        /// <summary>
        /// Analyze performance of different advanced query techniques.
        /// </summary>
        public async Task QueryPerformanceAnalysisAsync()
        {
            Console.WriteLine("\n" + new string('=', 80));
            Console.WriteLine("QUERY PERFORMANCE ANALYSIS");
            Console.WriteLine(new string('=', 80));

            var queries = new[]
            {
                ("Simple text", "azure machine learning", SearchQueryType.Simple),
                ("Wildcard", "develop*", SearchQueryType.Simple),
                ("Fuzzy", "machne~ learing~", SearchQueryType.Full),
                ("Complex", "python* AND tutorial", SearchQueryType.Full)
            };

            Console.WriteLine("\nPerformance Comparison:");
            Console.WriteLine(new string('-', 40));

            foreach (var (name, query, queryType) in queries)
            {
                try
                {
                    var startTime = DateTime.Now;

                    var searchOptions = new SearchOptions
                    {
                        QueryType = queryType,
                        Size = 10
                    };

                    var results = await _searchClient.SearchAsync<SearchDocument>(query, searchOptions);
                    var resultList = results.Value.GetResults().ToList();

                    var executionTime = DateTime.Now - startTime;

                    Console.WriteLine($"{name,-15} | {executionTime.TotalMilliseconds,6:F0}ms | {resultList.Count,3} results | {query}");
                }
                catch (RequestFailedException ex)
                {
                    Console.WriteLine($"{name,-15} | ERROR   | {ex.Message.Substring(0, Math.Min(50, ex.Message.Length))}...");
                }
            }

            Console.WriteLine("\nPerformance Tips:");
            Console.WriteLine("• Simple queries are fastest");
            Console.WriteLine("• Wildcard queries are moderately fast");
            Console.WriteLine("• Fuzzy queries add processing overhead");
            Console.WriteLine("• Complex combinations multiply overhead");
        }

        /// <summary>
        /// Main method to run all advanced query examples.
        /// </summary>
        public static async Task Main(string[] args)
        {
            Console.WriteLine("Azure AI Search - Advanced Queries Examples");
            Console.WriteLine(new string('=', 80));

            try
            {
                var advancedQueries = new AdvancedQueries();

                Console.WriteLine($"✅ Connected to search service");
                Console.WriteLine($"✅ Using configured index");

                // Run examples
                await advancedQueries.FieldBoostingExamplesAsync();
                await advancedQueries.FuzzySearchExamplesAsync();
                await advancedQueries.WildcardSearchExamplesAsync();
                await advancedQueries.ComplexQueryCombinationsAsync();
                await advancedQueries.QueryPerformanceAnalysisAsync();

                Console.WriteLine("\n" + new string('=', 80));
                Console.WriteLine("✅ All advanced query examples completed successfully!");
                Console.WriteLine(new string('=', 80));

                Console.WriteLine("\n📚 What you learned:");
                Console.WriteLine("• How to use field boosting to influence relevance");
                Console.WriteLine("• How to implement fuzzy search for typo tolerance");
                Console.WriteLine("• How to use wildcard patterns for flexible matching");
                Console.WriteLine("• How to combine multiple advanced techniques");
                Console.WriteLine("• How to analyze and optimize query performance");

                Console.WriteLine("\n🔗 Next steps:");
                Console.WriteLine("• Run 06_ErrorHandling.cs to learn robust query implementation");
                Console.WriteLine("• Experiment with different boosting values");
                Console.WriteLine("• Try complex query combinations with your data");
            }
            catch (InvalidOperationException ex)
            {
                Console.WriteLine($"❌ Configuration error: {ex.Message}");
                Console.WriteLine("\n🔧 Setup required:");
                Console.WriteLine("1. Create an appsettings.json file with your Azure AI Search credentials");
                Console.WriteLine("2. Or set environment variables for the search service configuration");
                Console.WriteLine("3. Ensure you have completed previous modules to create sample indexes");
                Environment.Exit(1);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"❌ Unexpected error: {ex.Message}");
                Environment.Exit(1);
            }
        }
    }
}

File: docs/beginner/module-04-simple-queries/code-samples/python/05_advanced_queries.py
#!/usr/bin/env python3
"""
Module 4: Simple Queries and Filters - Advanced Queries
=======================================================

This script demonstrates advanced query techniques in Azure AI Search using Python.
Learn about field boosting, fuzzy search, wildcards, and complex query patterns.

Prerequisites:
- Azure AI Search service configured
- Sample index with data (from previous modules)
- Environment variables set in .env file

Author: Azure AI Search Tutorial
"""

import os
import sys
from typing import List, Dict, Any
from azure.search.documents import SearchClient
from azure.core.credentials import AzureKeyCredential
from azure.core.exceptions import HttpResponseError
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

def create_search_client() -> SearchClient:
    """
    Create and return an Azure AI Search client.
    
    Returns:
        SearchClient: Configured search client
    """
    endpoint = os.getenv("AZURE_SEARCH_SERVICE_ENDPOINT")
    api_key = os.getenv("AZURE_SEARCH_API_KEY")
    index_name = os.getenv("AZURE_SEARCH_INDEX_NAME")
    
    if not all([endpoint, api_key, index_name]):
        raise ValueError("Missing required environment variables. Check your .env file.")
    
    return SearchClient(
        endpoint=endpoint,
        index_name=index_name,
        credential=AzureKeyCredential(api_key)
    )

def print_results(results: List[Dict[str, Any]], title: str, max_results: int = 5) -> None:
    """
    Print search results in a formatted way.
    
    Args:
        results: List of search result documents
        title: Title for the result set
        max_results: Maximum number of results to display
    """
    print(f"\n{'='*60}")
    print(f"{title}")
    print(f"{'='*60}")
    
    if not results:
        print("No results found.")
        return
    
    for i, result in enumerate(results[:max_results]):
        print(f"\n{i+1}. {result.get('title', 'No title')}")
        print(f"   Score: {result.get('@search.score', 'N/A'):.4f}")
        print(f"   Category: {result.get('category', 'N/A')}")
        print(f"   Rating: {result.get('rating', 'N/A')}")
        
        # Show content preview
        content = result.get('content', '')
        if content:
            preview = content[:100] + "..." if len(content) > 100 else content
            print(f"   Preview: {preview}")
    
    if len(results) > max_results:
        print(f"\n... and {len(results) - max_results} more results")

def field_boosting_examples(search_client: SearchClient) -> None:
    """
    Demonstrate field boosting to influence relevance scoring.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("FIELD BOOSTING EXAMPLES")
    print("="*80)
    
    # Example 1: No boosting (baseline)
    print("\n1. No Field Boosting (Baseline)")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="python tutorial",
            search_fields=["title", "content"],
            top=5
        ))
        print_results(results, "No boosting - equal weight for all fields")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 2: Boost title field
    print("\n2. Boost Title Field (3x weight)")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="python tutorial",
            search_fields=["title^3", "content"],  # Title weighted 3x
            top=5
        ))
        print_results(results, "Title boosted 3x - title matches score higher")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 3: Multiple field boosting
    print("\n3. Multiple Field Boosting")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="azure machine learning",
            search_fields=["title^5", "category^2", "content"],  # Title 5x, category 2x
            top=5
        ))
        print_results(results, "Title 5x, category 2x, content 1x")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 4: Extreme boosting comparison
    print("\n4. Extreme Boosting Comparison")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="tutorial",
            search_fields=["title^10", "content"],  # Extreme title boost
            top=5
        ))
        print_results(results, "Extreme title boosting (10x)")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def fuzzy_search_examples(search_client: SearchClient) -> None:
    """
    Demonstrate fuzzy search for handling typos and variations.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("FUZZY SEARCH EXAMPLES")
    print("="*80)
    
    # Example 1: Exact match (baseline)
    print("\n1. Exact Match (Baseline)")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="machine",
            query_type="simple",
            top=3
        ))
        print_results(results, "Exact match: 'machine'")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 2: Fuzzy search with typo
    print("\n2. Fuzzy Search with Typo")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="machne~",  # Typo: missing 'i'
            query_type="full",  # Requires full Lucene syntax
            top=3
        ))
        print_results(results, "Fuzzy search: 'machne~' (should match 'machine')")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 3: Fuzzy search with edit distance
    print("\n3. Fuzzy Search with Edit Distance")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="learing~2",  # Allow up to 2 character differences
            query_type="full",
            top=3
        ))
        print_results(results, "Fuzzy search: 'learing~2' (should match 'learning')")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 4: Multiple fuzzy terms
    print("\n4. Multiple Fuzzy Terms")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="machne~ learing~",  # Multiple fuzzy terms
            query_type="full",
            top=3
        ))
        print_results(results, "Multiple fuzzy: 'machne~ learing~'")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def wildcard_search_examples(search_client: SearchClient) -> None:
    """
    Demonstrate wildcard search patterns.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("WILDCARD SEARCH EXAMPLES")
    print("="*80)
    
    # Example 1: Suffix wildcard
    print("\n1. Suffix Wildcard")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="develop*",  # Matches develop, developer, development, etc.
            top=5
        ))
        print_results(results, "Suffix wildcard: 'develop*'")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 2: Prefix wildcard (requires full Lucene)
    print("\n2. Prefix Wildcard")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="*ing",  # Matches words ending in 'ing'
            query_type="full",
            top=5
        ))
        print_results(results, "Prefix wildcard: '*ing'")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 3: Middle wildcard
    print("\n3. Middle Wildcard")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="mach*ne",  # Matches machine, etc.
            query_type="full",
            top=5
        ))
        print_results(results, "Middle wildcard: 'mach*ne'")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 4: Single character wildcard
    print("\n4. Single Character Wildcard")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="te?t",  # Matches test, text, etc.
            query_type="full",
            top=5
        ))
        print_results(results, "Single char wildcard: 'te?t'")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def proximity_search_examples(search_client: SearchClient) -> None:
    """
    Demonstrate proximity search for terms near each other.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("PROXIMITY SEARCH EXAMPLES")
    print("="*80)
    
    # Example 1: Exact phrase
    print("\n1. Exact Phrase")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text='"machine learning"',  # Exact phrase
            top=3
        ))
        print_results(results, 'Exact phrase: "machine learning"')
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 2: Proximity search (within N words)
    print("\n2. Proximity Search (Within 5 Words)")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text='"machine learning"~5',  # Within 5 words of each other
            query_type="full",
            top=3
        ))
        print_results(results, 'Proximity: "machine learning"~5')
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 3: Looser proximity
    print("\n3. Looser Proximity (Within 10 Words)")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text='"azure cognitive"~10',
            query_type="full",
            top=3
        ))
        print_results(results, 'Proximity: "azure cognitive"~10')
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def regular_expression_examples(search_client: SearchClient) -> None:
    """
    Demonstrate regular expression search patterns.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("REGULAR EXPRESSION EXAMPLES")
    print("="*80)
    
    # Example 1: Simple regex pattern
    print("\n1. Simple Regex Pattern")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="/[Tt]utorial/",  # Matches Tutorial or tutorial
            query_type="full",
            top=3
        ))
        print_results(results, "Regex: /[Tt]utorial/")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 2: Number pattern
    print("\n2. Number Pattern")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="/[0-9]+/",  # Matches any number
            query_type="full",
            top=3
        ))
        print_results(results, "Regex: /[0-9]+/ (any number)")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 3: Version pattern
    print("\n3. Version Pattern")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="/[0-9]+\\.[0-9]+/",  # Matches version numbers like 3.8
            query_type="full",
            top=3
        ))
        print_results(results, "Regex: /[0-9]+\\.[0-9]+/ (version numbers)")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def complex_query_combinations(search_client: SearchClient) -> None:
    """
    Demonstrate complex combinations of advanced query features.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("COMPLEX QUERY COMBINATIONS")
    print("="*80)
    
    # Example 1: Boosting + Fuzzy + Filter
    print("\n1. Boosting + Fuzzy + Filter")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="machne~ learing~",  # Fuzzy search
            search_fields=["title^3", "content"],  # Field boosting
            filter="rating ge 3.0",  # Filter
            query_type="full",
            top=3
        ))
        print_results(results, "Fuzzy + boosting + filter combination")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 2: Wildcard + Proximity + Sorting
    print("\n2. Wildcard + Proximity + Sorting")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text='develop* AND "tutorial guide"~3',  # Wildcard + proximity
            order_by=["rating desc"],  # Sorting
            query_type="full",
            top=3
        ))
        print_results(results, "Wildcard + proximity + sorting")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")
    
    # Example 3: Multiple techniques with highlighting
    print("\n3. Multiple Techniques + Highlighting")
    print("-" * 40)
    
    try:
        results = list(search_client.search(
            search_text="python* OR java*",  # Wildcard OR
            search_fields=["title^2", "content"],  # Boosting
            filter="category eq 'Technology'",  # Filter
            highlight_fields=["title", "content"],  # Highlighting
            order_by=["@search.score desc"],  # Sort by relevance
            query_type="full",
            top=3
        ))
        
        print_results(results, "Complex combination with highlighting")
        
        # Show highlights
        for i, result in enumerate(results[:2]):
            if '@search.highlights' in result:
                print(f"\nHighlights for result {i+1}:")
                for field, highlights in result['@search.highlights'].items():
                    for highlight in highlights[:1]:
                        print(f"  {field}: {highlight}")
        
    except HttpResponseError as e:
        print(f"Search failed: {e.message}")

def query_performance_analysis(search_client: SearchClient) -> None:
    """
    Analyze performance of different advanced query techniques.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("QUERY PERFORMANCE ANALYSIS")
    print("="*80)
    
    import time
    
    queries = [
        ("Simple text", "azure machine learning", "simple"),
        ("Wildcard", "develop*", "simple"),
        ("Fuzzy", "machne~ learing~", "full"),
        ("Regex", "/[Tt]utorial/", "full"),
        ("Complex", 'python* AND "tutorial guide"~3', "full")
    ]
    
    print("\nPerformance Comparison:")
    print("-" * 40)
    
    for name, query, query_type in queries:
        try:
            start_time = time.time()
            
            results = list(search_client.search(
                search_text=query,
                query_type=query_type,
                top=10
            ))
            
            execution_time = time.time() - start_time
            
            print(f"{name:15} | {execution_time:6.3f}s | {len(results):3d} results | {query}")
            
        except HttpResponseError as e:
            print(f"{name:15} | ERROR   | {str(e)[:50]}...")
    
    print("\nPerformance Tips:")
    print("• Simple queries are fastest")
    print("• Wildcard queries are moderately fast")
    print("• Fuzzy queries add processing overhead")
    print("• Regex queries can be slow with complex patterns")
    print("• Complex combinations multiply overhead")

def advanced_scoring_examples(search_client: SearchClient) -> None:
    """
    Demonstrate advanced scoring and relevance techniques.
    
    Args:
        search_client: Azure AI Search client
    """
    print("\n" + "="*80)
    print("ADVANCED SCORING EXAMPLES")
    print("="*80)
    
    # Example 1: Score analysis with different query types
    print("\n1. Score Analysis by Query Type")
    print("-" * 40)
    
    query_types = [
        ("Simple", "azure tutorial", "simple"),
        ("Fuzzy", "azure~ tutorial~", "full"),
        ("Wildcard", "azure* tutorial*", "full")
    ]
    
    for name, query, query_type in query_types:
        try:
            results = list(search_client.search(
                search_text=query,
                query_type=query_type,
                top=3
            ))
            
            if results:
                scores = [r.get('@search.score', 0) for r in results]
                avg_score = sum(scores) / len(scores)
                print(f"\n{name} Query: '{query}'")
                print(f"  Average score: {avg_score:.4f}")
                print(f"  Score range: {min(scores):.4f} - {max(scores):.4f}")
            
        except HttpResponseError as e:
            print(f"{name} query failed: {e.message}")
    
    # Example 2: Boosting impact on scores
    print("\n2. Boosting Impact on Scores")
    print("-" * 40)
    
    try:
        # No boosting
        results_normal = list(search_client.search(
            search_text="python tutorial",
            search_fields=["title", "content"],
            top=3
        ))
        
        # With boosting
        results_boosted = list(search_client.search(
            search_text="python tutorial",
            search_fields=["title^5", "content"],
            top=3
        ))
        
        print("Score comparison (same query, different boosting):")
        print("\nNo boosting:")
        for i, result in enumerate(results_normal[:2]):
            print(f"  {i+1}. Score: {result.get('@search.score', 0):.4f} - {result.get('title', '')[:40]}...")
        
        print("\nWith title boosting (5x):")
        for i, result in enumerate(results_boosted[:2]):
            print(f"  {i+1}. Score: {result.get('@search.score', 0):.4f} - {result.get('title', '')[:40]}...")
        
    except HttpResponseError as e:
        print(f"Boosting comparison failed: {e.message}")

def main():
    """
    Main function to run all advanced query examples.
    """
    print("Azure AI Search - Advanced Queries Examples")
    print("=" * 80)
    
    try:
        # Create search client
        search_client = create_search_client()
        print(f"✅ Connected to search service: {os.getenv('AZURE_SEARCH_SERVICE_ENDPOINT')}")
        print(f"✅ Using index: {os.getenv('AZURE_SEARCH_INDEX_NAME')}")
        
        # Run examples
        field_boosting_examples(search_client)
        fuzzy_search_examples(search_client)
        wildcard_search_examples(search_client)
        proximity_search_examples(search_client)
        regular_expression_examples(search_client)
        complex_query_combinations(search_client)
        query_performance_analysis(search_client)
        advanced_scoring_examples(search_client)
        
        print("\n" + "="*80)
        print("✅ All advanced query examples completed successfully!")
        print("="*80)
        
        print("\n📚 What you learned:")
        print("• How to use field boosting to influence relevance")
        print("• How to implement fuzzy search for typo tolerance")
        print("• How to use wildcard patterns for flexible matching")
        print("• How to perform proximity and phrase searches")
        print("• How to use regular expressions in queries")
        print("• How to combine multiple advanced techniques")
        print("• How to analyze and optimize query performance")
        
        print("\n🔗 Next steps:")
        print("• Run 06_error_handling.py to learn robust query implementation")
        print("• Experiment with different boosting values")
        print("• Try complex query combinations with your data")
        print("• Monitor query performance in production")
        
    except ValueError as e:
        print(f"❌ Configuration error: {e}")
        print("\n🔧 Setup required:")
        print("1. Create a .env file with your Azure AI Search credentials")
        print("2. Ensure you have completed previous modules to create sample indexes")
        sys.exit(1)
        
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()